org.apache.fury.exception.DeserializationException: Deserialize failed, read objects are: [null]

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/fury/issues) and found no similar issues.


### Version

0.7.0-SNAPSHOT

### Component(s)

Java

### Minimal reproduce step

    @Test
    public void test() throws IOException {
        ThreadSafeFury fury = Fury.builder().withLanguage(Language.JAVA)
                //开启共享引用/循环引用支持
                .withRefTracking(true)
                .requireClassRegistration(false)
                //开启异步多线程编译
                .withStringCompressed(true)
                .withNumberCompressed(false)
                //开启类型前后兼容
                .withCompatibleMode(CompatibleMode.COMPATIBLE)
                .buildThreadSafeFury();
        ByteArrayOutputStream bas = new ByteArrayOutputStream();
        List list = new ArrayList<>();
        HashMap<String, String> map = new HashMap<>();
        for (int i = 0; i < 500000; i++) {
            list.add(i);
            map.put("key" + i, "value" + i);
        }
        fury.serialize(bas, list);
        fury.serialize(bas, map);
        fury.serialize(bas, list);
        bas.flush();

        InputStream bis = new ByteArrayInputStream(bas.toByteArray());
        FuryInputStream stream = of(bis);
        Object lobj1 = fury.deserialize(stream);
        Object lobj2 = fury.deserialize(stream);
        Object lobj3 = fury.deserialize(stream);
    }

    @Test
    public void test1() throws IOException {
        ThreadSafeFury fury = Fury.builder().withLanguage(Language.JAVA)
                //开启共享引用/循环引用支持
                .withRefTracking(true)
                .requireClassRegistration(false)
                //开启异步多线程编译
                .withStringCompressed(true)
                .withNumberCompressed(false)
                //开启类型前后兼容
                .withCompatibleMode(CompatibleMode.COMPATIBLE)
                .buildThreadSafeFury();
        ByteArrayOutputStream bas = new ByteArrayOutputStream();
        fury.serialize(bas, new HashMap[5000000]);
        fury.serialize(bas, new long[5000000]);
        fury.serialize(bas, new int[5000000]);
        bas.flush();

        InputStream bis = new ByteArrayInputStream(bas.toByteArray());
        FuryInputStream stream = of(bis);
        Object lobj1 = fury.deserialize(stream);
        Object lobj2 = fury.deserialize(stream);
        Object lobj3 = fury.deserialize(stream);
    }

### What did you expect to see?

no exception

### What did you see instead?

org.apache.fury.exception.DeserializationException: Deserialize failed, read objects are: [null]
	at org.apache.fury.util.ExceptionUtils.handleReadFailed(ExceptionUtils.java:63)
	at org.apache.fury.Fury.deserialize(Fury.java:793)
	at org.apache.fury.Fury.deserialize(Fury.java:812)
	at org.apache.fury.Fury.deserialize(Fury.java:805)
	at org.apache.fury.ThreadLocalFury.deserialize(ThreadLocalFury.java:160)
	at me.ele.roadnet.algo.odps.cli.FuryTest.test1(FuryTest.java:82)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)
Caused by: java.lang.NullPointerException
	at org.apache.fury.resolver.ClassResolver.getOrUpdateClassInfo(ClassResolver.java:1153)
	at org.apache.fury.resolver.ClassResolver.readClassInfoWithMetaShare(ClassResolver.java:1345)
	at org.apache.fury.resolver.ClassResolver.readClassInfo(ClassResolver.java:1603)
	at org.apache.fury.Fury.readRef(Fury.java:857)
	at org.apache.fury.Fury.deserialize(Fury.java:789)
	... 31 more

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/java/fury-core/src/main/java/org/apache/fury/Fury.java
+++ b/java/fury-core/src/main/java/org/apache/fury/Fury.java
@@ -345,7 +345,7 @@ public final class Fury implements BaseFury {
       writeData(buffer, classInfo, obj);
     }
     if (shareMeta) {
-      buffer.putInt32(startOffset, buffer.writerIndex());
+      buffer.putInt32(startOffset, buffer.writerIndex() - startOffset - 4);
       classResolver.writeClassDefs(buffer);
     }
   }
@@ -1064,7 +1064,7 @@ public final class Fury implements BaseFury {
           ClassInfo classInfo = classResolver.getOrUpdateClassInfo(obj.getClass());
           writeData(buffer, classInfo, obj);
         }
-        buffer.putInt32(startOffset, buffer.writerIndex());
+        buffer.putInt32(startOffset, buffer.writerIndex() - startOffset - 4);
         classResolver.writeClassDefs(buffer);
       } else {
         if (!refResolver.writeRefOrNull(buffer, obj)) {
@@ -1421,9 +1421,9 @@ public final class Fury implements BaseFury {
   }
 
   private void readClassDefs(MemoryBuffer buffer) {
-    int classDefOffset = buffer.readInt32();
+    int relativeClassDefOffset = buffer.readInt32();
     int readerIndex = buffer.readerIndex();
-    buffer.readerIndex(classDefOffset);
+    buffer.readerIndex(readerIndex + relativeClassDefOffset);
     classResolver.readClassDefs(buffer);
     classDefEndOffset = buffer.readerIndex();
     buffer.readerIndex(readerIndex);

--- a/java/fury-core/src/test/java/org/apache/fury/StreamTest.java
+++ b/java/fury-core/src/test/java/org/apache/fury/StreamTest.java
@@ -33,6 +33,7 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import org.apache.fury.config.CompatibleMode;
 import org.apache.fury.io.FuryInputStream;
 import org.apache.fury.io.FuryReadableChannel;
@@ -43,7 +44,7 @@ import org.apache.fury.test.bean.BeanA;
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
-public class StreamTest {
+public class StreamTest extends FuryTestBase {
 
   @Test
   public void testBufferStream() {
@@ -350,5 +351,31 @@ public class StreamTest {
     Assert.assertEquals(fury.deserialize(stream), list);
     Assert.assertEquals(fury.deserialize(stream), map);
     Assert.assertEquals(fury.deserialize(stream), list2);
+  }
+
+  @Test
+  public void testBigBufferStreamingMetaShared() throws IOException {
+    Fury fury = builder().withCompatibleMode(CompatibleMode.COMPATIBLE).build();
+    ByteArrayOutputStream bas = new ByteArrayOutputStream();
+    List<Integer> list = new ArrayList<>();
+    HashMap<String, String> map = new HashMap<>();
+    for (int i = 0; i < 5000; i++) {
+      list.add(i);
+      map.put("key" + i, "value" + i);
+    }
+    fury.serialize(bas, list);
+    fury.serialize(bas, map);
+    fury.serialize(bas, list);
+    fury.serialize(bas, new long[5000]);
+    fury.serialize(bas, new int[5000]);
+    bas.flush();
+
+    InputStream bis = new ByteArrayInputStream(bas.toByteArray());
+    FuryInputStream stream = of(bis);
+    assertEquals(fury.deserialize(stream), list);
+    assertEquals(fury.deserialize(stream), map);
+    assertEquals(fury.deserialize(stream), list);
+    assertEquals(fury.deserialize(stream), new long[5000]);
+    assertEquals(fury.deserialize(stream), new int[5000]);
   }
 }

